### PR TITLE
[#511] Use virtual host from EndpointContext in SNI extension.

### PR DESCRIPTION
**Pull request by sophokles73**
The ClientHandshaker by default will include an SNI extension in its
CLIENT_HELLO if the message that triggered the handshake is targeted at
a host name instead of a literal IP. However, SNI can be disabled using
the sniEnabled connector configuration property.

### CODE REVIEW

File: scandium-core/src/main/java/org/eclipse/californium/scandium/config/DtlsConnectorConfig.java [hunk 1]
```diff
@@ -145,6 +144,8 @@
 	 */
 	private Long autoResumptionTimeoutMillis;
 
+	private boolean sniEnabled = true;
```
Comment (boaks) [inline scandium-core/src/main/java/org/eclipse/californium/scandium/config/DtlsConnectorConfig.java]: With PR #336 we changed the design for this configuration to use references. 
This allows a upper layer to provide it's own defaults by checking, if the value was set, or a default should be used (that's the null case). I would prefer, if this `sniEnable` gets changed to `Boolean`.

Comment (sophokles73) [reply scandium-core/src/main/java/org/eclipse/californium/scandium/config/DtlsConnectorConfig.java]: So, the idea would be to also add a getter for this property in the builder and then set the property to its default (true) in the *build()* method if it hasn't been set explicitly yet, right?

Comment (boaks) [reply scandium-core/src/main/java/org/eclipse/californium/scandium/config/DtlsConnectorConfig.java]: yes.

**Comment by boaks** [comment]
Would it be possible to add

			builder.setSniEnabled(false);

to demo-apps/cf-secure/.../SecureClient.java?

File: scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/DTLSSession.java [hunk 2]
```diff
@@ -243,6 +262,37 @@ void setSessionIdentifier(SessionId sessionIdentifier) {
...
+	public boolean isSniSupported() {
+		return peerSupportsSni;
+	}
+
+	void setSniSupported(boolean flag) {
+		this.peerSupportsSni = flag;
+	}
+
 	public DtlsEndpointContext getConnectionWriteContext() {
 		return new DtlsEndpointContext(peer, peerIdentity, sessionIdentifier.toString(),
```
Comment (boaks) [inline scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/DTLSSession.java]: Please add the SNI/virtualHost to the context.

Comment (sophokles73) [reply scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/DTLSSession.java]: That will be in the next step ...

Comment (boaks) [reply scandium-core/src/main/java/org/eclipse/californium/scandium/dtls/DTLSSession.java]: OK, then please also add this to the read-context.

File: scandium-core/src/main/java/org/eclipse/californium/scandium/config/DtlsConnectorConfig.java [hunk 3]
```diff
@@ -205,6 +206,19 @@ public Boolean isAddressReuseEnabled() {
...
+	/**
+	 * Checks whether the connector should support the use of the TLS
+	 * <a href="https://tools.ietf.org/html/rfc6066#section-3">
+	 * Server Name Indication extension</a> in the DTLS handshake.
+	 * <p>
+	 * The default value of this property is {@code true}.
+	 * 
+	 * @return {@code true} if SNI should be used.
+	 */
+	public boolean isSniEnabled() {
```
Comment (boaks) [inline scandium-core/src/main/java/org/eclipse/californium/scandium/config/DtlsConnectorConfig.java]: The getter is also intended to return a "Boolean" reference, so someone can test for null to check, if the default is intended to be used. You may check leshan builders to see, how it is intended to work.

Comment (sophokles73) [reply scandium-core/src/main/java/org/eclipse/californium/scandium/config/DtlsConnectorConfig.java]: sure, oversaw this ...

**Comment by boaks** [comment]
LGTM
